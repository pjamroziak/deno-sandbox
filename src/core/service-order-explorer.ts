import { Constructor } from "./types.ts";

export class ServiceOrderExplorer {
  private readonly graph = new Map<string, string[]>();

  search(): string[] {
    const visited = new Set<string>();
    const stack: string[] = [];

    if (this.graph.size === 0) return [];

    for (const [node] of this.graph) {
      this.explore(node, visited, stack);
    }

    return stack.reverse();
  }

  addService(service: Constructor, dependencies: Constructor[] = []) {
    this.addNode(service.name);

    const dependenciesNames = dependencies.map((dependency) => dependency.name);
    for (const dependency of dependenciesNames) {
      this.addNode(dependency);
    }

    this.graph.get(service.name)!.push(...dependenciesNames);

    return this;
  }

  private explore(
    node: string,
    visited: Set<string>,
    stack: string[],
  ): void {
    if (!visited.has(node)) visited.add(node);

    const dependencies = this.graph.get(node) ?? [];
    for (const child of dependencies) {
      if (!visited.has(child)) this.explore(child, visited, stack);
    }

    if (!stack.includes(node)) stack.unshift(node);
  }

  private addNode(node: string) {
    if (!this.graph.has(node)) {
      this.graph.set(node, []);
    }
  }
}
