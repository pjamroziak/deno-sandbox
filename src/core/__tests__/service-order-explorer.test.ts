import { ServiceOrderExplorer } from "../service-order-explorer.ts";
import { beforeEach, describe, it } from "@std/testing/bdd";
import { assertSnapshot } from "@std/testing/snapshot";

describe(ServiceOrderExplorer.name, () => {
  let explorer: ServiceOrderExplorer;
  beforeEach(() => {
    explorer = new ServiceOrderExplorer();
  });

  it("should return array of all necessary services, ordered by construction order", async (t) => {
    explorer
      .addService("a", ["b", "c"])
      .addService("b", ["c"])
      .addService("c", ["d"])
      .addService("d");

    await assertSnapshot(t, explorer.search());
  });

  it("should return array with one service", async (t) => {
    explorer
      .addService("d");

    await assertSnapshot(t, explorer.search());
  });

  it("should return array of all necessary services, ordered by construction order when service has circular dependency", async (t) => {
    explorer
      .addService("a", ["b", "c", "a"])
      .addService("b", ["c"])
      .addService("c", ["d"])
      .addService("d");

    await assertSnapshot(t, explorer.search());
  });
});
