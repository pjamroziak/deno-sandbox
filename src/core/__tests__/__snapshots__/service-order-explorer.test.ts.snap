export const snapshot = {};

snapshot[`ServiceOrderExplorer > should return array of all necessary services, ordered by construction order 1`] = `
[
  "d",
  "c",
  "b",
  "a",
]
`;

snapshot[`ServiceOrderExplorer > should return array with one service 1`] = `
[
  "d",
]
`;

snapshot[`ServiceOrderExplorer > should return array of all necessary services, ordered by construction order when service has circular dependency 1`] = `
[
  "d",
  "c",
  "b",
  "a",
]
`;
